pubtools/_content_gateway/push_cgw.py

```diff
@@ -1,2 +1,196 @@
+        file_mapping = {(product_name, product_code, version_name, data.get('downloadURL')): (
+            product_id, version_id, data.get('id')) for data in all_files}
+        pvf_id = file_mapping.get(
+            (product_name, product_code, version_name,
+             file_item.get('metadata').get('downloadURL'))) if file_mapping else (None, None, None)
+        return pvf_id
+
+    def process_product(self, item):
+        try:
+            product_name = item.get('metadata').get('name')
+            product_code = item.get('metadata').get('productCode')
+            LOG.debug(f"Fetching for product_id of product name- {product_name} and product code- {product_code}")
+            product_id = self.__get_product_id(product_name, product_code)
+
+            if item.get('state') == 'present':
+                if not product_id:
+                    LOG.info("No previous entries found for the given product name and product code")
+                    LOG.info("Creating product entry for the given product metadata")
+                    product_id = self.cgw_client.create_product(item.get('metadata'))
+                    # A new product is created so updating product mapping
+                    cgw_data = self.cgw_client.get_products()
+                    self.product_mapping = {(data['name'], data['productCode']): data['id'] for data in cgw_data}
+                    LOG.debug(f"Created a new product with product_id:- {product_id}")
+                    return
+                LOG.info(f"Product record found with product_id:- {product_id}")
+                item.get('metadata')['id'] = product_id
+                LOG.info("Updating product metadata")
+                self.cgw_client.update_product(item.get('metadata'))
+            elif item.get('state') == 'absent' and product_id:
+                LOG.info(f"Deleting existing product records for product_id:- {product_id}")
+                self.cgw_client.delete_product(product_id)
+                LOG.info("Product record deleted!")
+            else:
+                LOG.warning("Invalid state or metadata!")
+        except Exception as err:
+            LOG.exception(err)
+
+    def process_version(self, item):
+        try:
+            product_name = item.get('metadata')['productName']
+            product_code = item.get('metadata')['productCode']
+            version_name = item.get('metadata')['versionName']
+            product_id = self.__get_product_id(product_name, product_code)
+            LOG.debug(f"Fetching the product and version id of "
+                      f"product name- {product_name} and version name- {version_name}")
+            version_id = self.__get_version_id(product_name, product_code, version_name)
+
+            if item.get('state') == 'present':
+                item.get('metadata')['productId'] = product_id
+                del item.get('metadata')['productName'], item.get('metadata')['productCode']
```

Comment:
Why?

Reply:
I am deleting this productName and productCode from the metadata to make the request body compatible with the required request body. We don't need these two during the API call else it will cause `400 Bad Request, The request cannot be fulfilled due to bad syntax.`